indexer/SymbolFormatter.cc

```diff
@@ -40,44 +41,194 @@ std::string_view SymbolFormatter::getMacroSymbol(clang::SourceLocation defLoc) {
+    return this->getSymbolCached(
+        decl, [&]() -> std::string { return "c . todo-pkg todo-version "; });
+  }
+  // TODO: Handle all cases of DeclContext here:
+  // Done
+  // - TranslationUnitDecl
+  // - ExternCContext
+  // - NamespaceDecl
+  // - TagDecl
+  // Pending:
+  // - OMPDeclareReductionDecl
+  // - OMPDeclareMapperDecl
+  // - FunctionDecl
+  // - ObjCMethodDecl
+  // - ObjCContainerDecl
+  // - LinkageSpecDecl
+  // - ExportDecl
+  // - BlockDecl
+  // - CapturedDecl
+  return std::nullopt;
+}
+
+std::optional<std::string_view>
+SymbolFormatter::getTagSymbol(const clang::TagDecl *tagDecl) {
+  return this->getSymbolCached(tagDecl, [&]() -> std::string {
+    auto optContextSymbol = this->getContextSymbol(tagDecl->getDeclContext());
+    if (!optContextSymbol.has_value()) {
+      return {};
+    }
+    auto contextSymbol = optContextSymbol.value();
+    if (tagDecl->getDeclName().isEmpty()) {
+      auto definitionTagDecl = tagDecl->getDefinition();
+      ENFORCE(definitionTagDecl, "can't forward-declare an anonymous type");
+      auto defLoc = definitionTagDecl->getLocation();
+      auto defPLoc = this->sourceManager.getPresumedLoc(
+          defLoc.isMacroID() ? this->sourceManager.getSpellingLoc(defLoc)
+                             : defLoc);
+      // TODO: Add test where definition is inside macro
+      ENFORCE(defPLoc.isValid());
+      auto defFileId = defPLoc.getFileID();
+      ENFORCE(defFileId.isValid());
+      auto counter = this->anonymousTypeCounters[{defFileId}]++;
+      if (!llvm::isa<clang::TagDecl>(definitionTagDecl->getDeclContext())) {
+        // If the anonymous type is inside a namespace, then we know the decl
+        // context chain only consists of namespaces (types cannot contain
+        // namespaces). So include the file hash too, to avoid collisions across
+        // files putting anonymous types into the same namespace. For example,
+        //   // A.h
+        //   namespace z { struct { void f() {} } x; }
+        //   // B.h
+        //   namespace z { struct { int f() { return 0; } } y; }
+        //
+        // If we don't include the hash, the anonymous structs will end up with
+        // the same symbol name.
+        if (auto optRelativePath = this->getCanonicalPath(defFileId)) {
+          HashValue hashValue{0};
+          auto sv = optRelativePath->asStringView();
+          hashValue.mix(reinterpret_cast<const uint8_t *>(sv.data()),
+                        sv.size());
+          return fmt::format("{}$anontype_{}_{}#", contextSymbol,
+                             hashValue.rawValue, counter);
+        }
+      }
+      return fmt::format("{}$anontype_{}#", contextSymbol, counter);
+    }
+    llvm::raw_string_ostream os(this->scratchBuffer);
+    static_cast<const clang::NamedDecl *>(tagDecl)->printName(os);
+    std::string out{fmt::format("{}{}#", contextSymbol, this->scratchBuffer)};
+    this->scratchBuffer.clear();
+    return out;
```

Comment:
Maybe flip the cases, and handle this early, since the other branch is long because of the comment